velox/functions/prestosql/fuzzer/CastFuzzerTest.cpp

```diff
@@ -0,0 +1,271 @@
+
+      for (auto i = 0; i < veloxCastVector->size(); i++) {
+        auto matchType = compareVarchars(prestoCastVector, veloxCastVector, i);
+
+        if (matchType == MatchType::EXACT_MATCH) {
+          exactMatch++;
+        } else if (matchType == MatchType::LEAST_SIGNIFICANT_DIGIT_OFF) {
+          leastSignificantDigitOff++;
+        } else {
+          VELOX_FAIL(fmt::format(
+              "Cast conversion not matching for: {} to {}",
+              prestoCastVector->valueAt(i),
+              veloxCastVector->valueAt(i)));
+        }
+      }
+
+      LOG(INFO) << fmt::format(
+          "Results EXACT : {}, LEAST SIGNIFICANT: {}",
+          exactMatch,
+          leastSignificantDigitOff);
+
+      seed(rng_());
+      ++iteration;
+    }
+  };
+
+ private:
+  enum class MatchType {
+    EXACT_MATCH = 1,
+    LEAST_SIGNIFICANT_DIGIT_OFF,
+    NOT_A_MATCH
+  };
+
+  /// Compare's two StringView vectors at index i. The StringView vectors are
+  /// results of conversion of a floating point type to string. Returns
+  /// EXACT_MATCH if both the strings are eactly matching.
+  /// LEAST_SIGNIFICANT_DIGIT_OFF if they are off in the least siginficant digit
+  /// NOT_A_MATCH if exponents or any other part of mantissa doesnt match.
+  MatchType compareVarchars(
+      const FlatVector<StringView>* expected,
+      const FlatVector<StringView>* actual,
+      size_t i) {
+    VELOX_CHECK_EQ(expected->size(), actual->size());
+    VELOX_CHECK_LT(i, expected->size());
+
+    if (expected->isNullAt(i) && actual->isNullAt(i)) {
+      return MatchType::EXACT_MATCH;
+    } else if (expected->isNullAt(i) || actual->isNullAt(i)) {
+      return MatchType::NOT_A_MATCH;
+    }
+
+    auto expectedString = expected->valueAt(i).getString();
+    auto actualString = actual->valueAt(i).getString();
+
+    if (expectedString == actualString) {
+      return MatchType::EXACT_MATCH;
+    }
+
+    auto splitIntoMantissaExponent = [](std::string& val) {
+      auto pos = val.find('E');
+      return std::tuple(val.substr(0, pos - 1), val.substr(pos + 1));
+    };
+
+    auto [expectedMantissa, expectedExponent] =
+        splitIntoMantissaExponent(expectedString);
+    auto [actualMantissa, actualExponent] =
+        splitIntoMantissaExponent(actualString);
+
+    // Exponents should always be equal
+    VELOX_CHECK_EQ(actualExponent, expectedExponent);
+
+    if (expectedMantissa == actualMantissa) {
+      return MatchType::EXACT_MATCH;
+    }
+
+    // Handle case when the last significant digit is off or missing
+    auto smallestLength = expectedMantissa.length() < actualMantissa.length()
+        ? expectedMantissa.length()
+        : actualMantissa.length();
+
+    if (expectedMantissa.substr(0, smallestLength) ==
+        actualMantissa.substr(0, smallestLength)) {
+      return MatchType::LEAST_SIGNIFICANT_DIGIT_OFF;
+    }
+
+    // Note we dont handle cases like ..5999 and ..6000 where last n digits get
+    // rounded off
```

Comment:
For a case like 0.59 and 0.6 it shouldnt hit the exponential part of the code, and would instead fall in the > 10 ^-3 to 10^7 boundary and so should return precisely the same result. 
In general for exponentials like this: 0.5950000000000000E1 or 0.60000000..E1 this will return that they dont match (and I would agree, this looks like a very large error ). 
Typically in my testing the digits dont match only at the 15,16 or 17th digit and these occur roughly only 20-30% of time (at that precision).